### ISSUE TITLE
Size increase

### ISSUE THREAD

**Issue by annevk** [tc39/proposal-type-annotations#35]
While the FAQ suggests this might remove a roadblock to shipping JS as-is, it doesn't really address the obvious counterargument. That the increase in file size is not good for bandwidth/performance and therefore compilation will continue to be needed. It would be great if it could spell out what I'm missing.

**Comment by Talent30** [comment]
https://github.com/tc39/proposal-binary-ast
Time to pus this proposal

**Comment by ljharb** [comment]
> i n environments like node and deno, the size of the code base is not important.

i strongly disagree with this claim; adding cost to every loaded and parsed file is a serious concern. v8 even has a threshold for function optimization that existing comments can cause to be surpassed, causing a serious performance hit.

**Comment by retripcon** [comment]
I also see the increased bundle size as a strong counter argument. What you will learn very quickly in today's web development is that bundle size absolutely matters, especially in bigger applications. By leaving the types in the source code, this indirectly encourages bigger bundle sizes because the transpilation is no longer required. The overhead of TS transpilation is moved from a one time effort for developers as increased bundle size and additional work for the JS engines to every users browser which should be a big red flag! For production scenarios this issue is quite obvious. By saying that minification is required for production anyways you can also add the transpilation step to your pipeline as you already have a pipeline running.

In development I think there are tons of tools out there which already have fast and realiable live reloading capabilities so load time in development is also not really an argument for me. Until I have switched from my IDE to my browser the live reload has already done its job and is presenting the latest version. 

Also the barrier for new developers is not really lowered that much. For production a pipeline with minification etc. should be run anyway. Adding a new step for TS support is mostly a one liner and takes you like 1 minute of googling. The same applies for the development process. Most IDEs already support live reloading and transpilation for TS and have integrated this quite well. 

To be clear: I am not against typing in general. It's actually quite the opposite and I use TS in all of my projects! But I don't really see any benefit in having optional types which are totally ignored at runtime if this increases bundle size by quite a lot.

**Comment by ljharb** [comment]
In practice, i don’t think the bundle size impact is an issue for larger web apps, because a build tool will virtually always be required for any application of substance for a plethora of other reasons, but for smaller projects and for server applications, this could be a surprising footgun.

**Comment by VitorLuizC** [comment]
I think we're comparing different things. We should compare size of JSDoc vs. types as comments, instead of untyped vs. typed size

**Comment by ljharb** [comment]
That’s only the comparison if the target/primary audience of this proposal is jsdoc users, which i suspect is a very small population. If instead, it’s people who would otherwise use untyped code, then the latter is a better comparison.

**Comment by ide** [comment]
This is a data point to get a ballpark idea of the size difference between distributing typed vs. untyped code without other desugaring and minification. I compared all of the TypeScript source files for a real-world Node server with their counterparts that had types removed with SWC and code reformatted with Prettier. The SWC configuration left the code otherwise as untouched as possible, preserving ES imports, comments, etc. 

| Source type   | Total size | Total size, compressed with brotli -9 |
|---------------|------------|---------------------------------------|
| TypeScript    | 2.426 MiB  | 0.334 MiB                             |
| Types removed | 2.027 MiB  | 0.288 MiB                             |

This is just one rough data point. The formatting between the two sets of files is slightly different despite using the same Prettier configuration for both. The grammar in this TC39 proposal is slightly different than TypeScript's.

I thought a Node server made for a good example since empirically I've found Node servers have lighter build pipelines than websites, if at all. It's plausible for a Node server's build step just to be tsc or other type-desugarer like esbuild, Sucrase, or SWC.

**Comment by benjamingr** [comment]
Good data point @ide !

I certainly hope to ship my Node.js servers and tools with type-comments in the future since the size impact isn't significant for my use case (small'ish servers for micro-services typically), parsing time usually isn't very significant (they are long running) and it would save a lot of work (basically reduce the work CI does to "run `npx tsc --noEmit`").

**Comment by tomByrer** [comment]
+1 increased JS size.
People who want to develop JS-target with types can get that right now: Typescript & great IDE support.  There are also other typed transpile-into-JS like [Nim](https://nim-lang.org/features.html) & [V-lang](https://vlang.io/).  If you want to deploy typed JS comments for some reason, there is [JSDoc](https://jsdoc.app/tags-type.html).

Bandwidth scarcity is a real thing, eg throttling cell phone plans after exceeding 4g/5g caps, areas where information is desperately needed but not readily accessible (current events in Eastern Europe, Pacific Rim, etc).   While the extra CPU & memory usage will be small, it [will NOT be zero](https://github.com/giltayar/proposal-types-as-comments/issues/49#issuecomment-1065931759).

Bad enough that JS is taking more & more resources; do we really **need** to encourage extra wasted 'comments'?

**Comment by tomByrer** [comment]
from @orta in my other thread:
> if you would minimize your code for deployment then that same process would strip types.

If you had a build step, then why not just use TypeScript?!?  (50% of JS market share AFAIK)

True, most folks do minify somewhere, [~15% delivered JS is not](https://youtu.be/JvJ0v5OohNg?list=PLakykuPxo3cjjfDJTiQ6Ew2FsorWuq5ih).  Sometimes when author does not, instead relies on CDNs/servers for the minfication step. 
 So now that's more bandwidth/CPU/energy CDNs/servers have to spend, where TS would have solved the typing issue.

**Comment by noelmace** [comment]
> « That’s only the comparison if the target/primary audience of this proposal is jsdoc users, which i suspect is a very small population. If instead, it’s people who would otherwise use untyped code, then the latter is a better comparison. » - @ljharb https://github.com/giltayar/proposal-types-as-comments/issues/35#issuecomment-1064210067

Good point. This is one of the reasons why I find it difficult to see a real interest in this proposal as:
1. JSDoc users is a very small population indeed (I'm part of it, but try to avoid to be biased :sweat_smile:)
2. by definition, people who use untyped code don't use related transpilers and therefore don't need this proposal

**Comment by alshdavid** [comment]
In fairness to the development argument, there have been times where I have created an `index.html` file on my desktop, scratched together a POC for something I was working on and really missed the intellisense you get with types.

I have also scratched together the most throw-away admin dashboards (I am not proud of this) with Alpine/PetiteVue in a script tag and everything else - I am not sure if this proposal would enable my laziness :laughing:. 

But for anything that will require longer than 5 minutes of effort, I practically have a keyboard macro for setting up the whole TypeScript compiler enchilada.

The proposal is opt-in so I don't see it as being a bad addition to the spec, it might even help people with legacy codebases make the transition to TypeScript easier.

My biggest concern would be that it takes resources away from finalising and implementing Web Assembly (mostly from Apple who drag their feet on everything). With file size being as big a deal as it is, I am surprised more people aren't piling onto the discussions there to help it become a compile target for TypeScript (totally doable espc given GC and DOM access slated for the future)
